### PR TITLE
Add `required` field to `ProgramQuestionDefinition` 

### PR DESCRIPTION
**Pull request by kendrickt**
### Description
This change introduces the `optional` field and allows admins to configure it for each program question definition. The field **does not affect anything** yet.

* Add `optional` field to `ProgramQuestionDefinition`. Old program question definitions will default to required.
* Program service method added to toggle `optional` for program question definitions
* Controller method added to toggle `optional` for program question definitions. 
  * changed controller method to delete program question definitions to set `question definition id` in the route, and not in the form. This controller method doesn't need a form anymore.
* Added optional/required toggle to program block edit views for each program question definition
  * small change to this view to have an explicit `delete program question definition` button

### Screenshots

First question is required, second question is optional.

optional required question toggle 1

After toggling both questions, the first question is optional and second question is required.

optional required question toggle 2

### Checklist
- [x] Created tests which fail without the change (if possible)

### Issue(s)
Progress towards #1307

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java [hunk 1]
```diff
@@ -18,6 +18,15 @@
   @JsonProperty("id")
   public abstract long id();
 
+  /**
+   * True if this program question definition is required. Otherwise it is optional.
+   *
+   * <p>This field was added in June. Program question definitions created before this field will
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java:24]: super nit:

```suggestion
   * <p>This field was added in June 2021. Program question definitions created before this field will
```

File: universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java [hunk 2]
```diff
@@ -31,12 +40,31 @@ public boolean hasQuestionDefinition() {
...
+
+  /** Return a program question definition with the {@link QuestionDefinition} set. */
+  public ProgramQuestionDefinition setQuestionDefinition(QuestionDefinition questionDefinition) {
     return new AutoValue_ProgramQuestionDefinition(
-        questionDefinition.getId(), Optional.of(questionDefinition));
+        questionDefinition.getId(), required(), Optional.of(questionDefinition));
+  }
+
+  /** Return a program question definition with a new required setting. */
+  public ProgramQuestionDefinition setRequired(boolean required) {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java:67]: question (non-blocking): Should we use a builder here instead?

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java [hunk 3]
```diff
@@ -243,6 +243,23 @@ ProgramDefinition setBlockPredicate(
   ProgramDefinition deleteBlock(long programId, long blockDefinitionId)
       throws ProgramNotFoundException, ProgramNeedsABlockException;
 
+  /**
+   * Set a program question definition to required or optional. If the question definition ID is not
+   * present in the program's block, then nothing is changed.
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramService.java:248]: question (non-blocking): Why not throw if the question isn't present?

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 4]
```diff
@@ -411,6 +411,26 @@ public ProgramDefinition setBlockPredicate(
...
+  @Transactional
+  public ProgramDefinition setProgramQuestionDefinitionRequired(
+      long programId, long blockDefinitionId, long questionDefinitionId, boolean required)
+      throws ProgramNotFoundException, ProgramBlockDefinitionNotFoundException {
+    ProgramDefinition programDefinition = getProgramDefinition(programId);
+    BlockDefinition blockDefinition = programDefinition.getBlockDefinition(blockDefinitionId);
+
+    ImmutableList<ProgramQuestionDefinition> programQuestionDefinitions =
+        blockDefinition.programQuestionDefinitions().stream()
+            .map(pqd -> pqd.id() == questionDefinitionId ? pqd.setRequired(required) : pqd)
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:424]: praise: clever use of `map` :)

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java [hunk 5]
```diff
@@ -321,63 +321,128 @@ private ContainerTag blockEditPanel(
...
+                Styles.FLEX,
+                Styles.GAP_1,
+                Styles.ITEMS_CENTER,
+                isRequired ? Styles.TEXT_BLACK : Styles.TEXT_GRAY_400,
+                Styles.FONT_MEDIUM,
+                Styles.BG_TRANSPARENT,
+                Styles.ROUNDED_FULL,
+                StyleUtils.hover(Styles.BG_GRAY_400, Styles.TEXT_GRAY_300))
+            .withType("submit")
+            .with(p("required").withClasses("hover-group:text-white"))
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java:385]: question (non-blocking): I think in the mocks this read "optional", does it matter?

**Review by carolinedanzi — APPROVED** [review]
This looks fantastic! One question I have is whether it's possible to test loading existing questions - do they parse correctly? Is it possible to add a unit test for that or something?

File: universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java (hunk 2 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java:67]: I'd rather not. It isn't so complex that it needs a builder yet.

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java (hunk 5 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/views/admin/programs/ProgramBlockEditView.java:385]: it doesn't. we can flip it on the front end once we do more styling. 

my two cents: it seems like `required` should be what is shown, if that is the special case. If questions are default optional, then the toggle should be to turn on required.

File: universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java [hunk 6]
```diff
@@ -18,6 +18,15 @@
   @JsonProperty("id")
   public abstract long id();
 
+  /**
+   * True if this program question definition is required. Otherwise it is optional.
+   *
+   * <p>This field was added in June. Program question definitions created before this field will
+   * default to optional (false).
```
[the remark this answers is not in the document]
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramQuestionDefinition.java:25]: Just checked with Neel. Sounds like we want `required` to be default. I'll make the change!

**Review by AWBaumann — APPROVED** [review]

**Pull request merged**